server/container/util-java8/src/main/java/org/apache/james/util/streams/JamesCollectors.java

```diff
@@ -19,26 +19,63 @@
 
 package org.apache.james.util.streams;
 
-import java.util.List;
-import java.util.Map;
+import java.util.Collection;
+import java.util.Set;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
 import java.util.function.Function;
+import java.util.function.Supplier;
 import java.util.stream.Collector;
-import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
 
 public class JamesCollectors {
-    public static <D> Collector<D, ?, Map<Integer, List<D>>> chunker(int chunkSize) {
-        Preconditions.checkArgument(chunkSize > 0, "ChunkSize should be strictly positive");
-        AtomicInteger counter = new AtomicInteger(-1);
-        return Collectors.groupingBy(x -> counter.incrementAndGet() / chunkSize);
+    public static <D> Collector<D, ?, Stream<Collection<D>>> chunker(int chunkSize) {
+        return new ChunkCollector<>(chunkSize);
     }
 
-    public static <D> Function<Stream<D>, Stream<List<D>>> chunk(int chunkSize) {
-        return stream -> stream.collect(chunker(chunkSize))
-            .values()
-            .stream();
+    public static class ChunkCollector<D> implements Collector<D, Multimap<Integer, D>, Stream<Collection<D>>> {
+        private final int chunkSize;
+        private final AtomicInteger counter;
+
+        private ChunkCollector(int chunkSize) {
+            Preconditions.checkArgument(chunkSize > 0, "ChunkSize should be strictly positive");
+            this.chunkSize = chunkSize;
+            this.counter = new AtomicInteger(-1);
+        }
+
+        @Override
+        public Supplier<Multimap<Integer, D>> supplier() {
+            return () -> Multimaps.synchronizedListMultimap(ArrayListMultimap.<Integer, D>create());
```

Comment:
a list of list should really be enough : you create a list and put elements until size == chunksize and then create a new one, etc. Append all list to a main list and you're done, right ?